[GTK] Enhanced <input type=color>: support alpha picker

lukewarlow opened the pull request:
#### 04790d591e76c50e57759ee3af70f58f0082fced
<pre>
[GTK] Enhanced &lt;input type=color&gt;: support alpha picker
<a href="https://bugs.webkit.org/show_bug.cgi?id=290266">https://bugs.webkit.org/show_bug.cgi?id=290266</a>

Reviewed by NOBODY (OOPS!).

Enable the GTK color picker to show its alpha slider when
the enhancements preference is enabled and the alpha attribute is used.

Note previously GTK color pickers always showed an alpha slider which never affected
the actual value, this patch also fixes this behaviour.

This adds a new API method webkit_color_chooser_request_get_allow_alpha
so that embedders can choose to update their UI accordingly.

* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/gtk/accessibility/color-well-expected.txt: Added.
* LayoutTests/platform/wpe/TestExpectations:
* Source/WebCore/accessibility/atspi/AccessibilityObjectTextAtspi.cpp:
(WebCore::AccessibilityObjectAtspi::text const):
* Source/WebKit/UIProcess/API/gtk/PageClientImpl.cpp:
(WebKit::PageClientImpl::createColorPicker):
* Source/WebKit/UIProcess/API/gtk/WebKitColorChooser.cpp:
(WebKit::WebKitColorChooser::create):
(WebKit::WebKitColorChooser::WebKitColorChooser):
* Source/WebKit/UIProcess/API/gtk/WebKitColorChooser.h:
* Source/WebKit/UIProcess/API/gtk/WebKitColorChooserRequest.cpp:
(webkitColorChooserRequestGetProperty):
(webkit_color_chooser_request_class_init):
(webkit_color_chooser_request_get_allow_alpha):
(webkitColorChooserRequestCreate):
* Source/WebKit/UIProcess/API/gtk/WebKitColorChooserRequest.h.in:
* Source/WebKit/UIProcess/gtk/WebColorPickerGtk.cpp:
(WebKit::WebColorPickerGtk::create):
(WebKit::WebColorPickerGtk::WebColorPickerGtk):
(WebKit::WebColorPickerGtk::showColorPicker):
* Source/WebKit/UIProcess/gtk/WebColorPickerGtk.h:
(WebKit::WebColorPickerGtk::supportsAlpha const):
* Tools/MiniBrowser/gtk/BrowserTab.c:
(runColorChooserCallback):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/04790d591e76c50e57759ee3af70f58f0082fced

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/96899 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/16513 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/6705 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/101973 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/47420 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/98944 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/16809 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/24960 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73833 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31041 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/99902 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/12669 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/87665 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54167 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/12429 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/5486 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/46747 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/82520 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/5567 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/103996 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/23967 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/17494 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82879 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/24220 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/83739 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82272 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/26938 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/4495 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/17470 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/23929 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/29084 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/23588 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/27068 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/25329 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

philn commented on Source/WebCore/accessibility/atspi/AccessibilityObjectTextAtspi.cpp:
```diff
@@ -277,7 +277,8 @@ String AccessibilityObjectAtspi::text() const
     if (m_coreObject->roleValue() == AccessibilityRole::ColorWell) {
         auto color = convertColor<SRGBA<float>>(m_coreObject->colorValue()).resolved();
-        GUniquePtr<char> colorString(g_strdup_printf("rgb %7.5f %7.5f %7.5f 1", color.red, color.green, color.blue));
+        const char* alphaFormat = (color.alpha == 1.0f) ? "1" : g_strdup_printf("%7.5f", color.alpha);
```
The string returned by `g_strdup_printf()` is leaked here.

lukewarlow replied:
Apologies I don't really understand what this means?

philn replied:
that function allocates a new string, so it needs to be released after usage, usually with g_free(). We have `GUniquePtr<char>()` that can manage this for you.

Maybe this can be reworked using `StringPrintStream`?

philn replied:
https://github.com/philn/WebKit/commit/415cd39950016810688f3cc4fbf862a28113a586 maybe? (untested)

lukewarlow replied:
I'll give that a go tomorrow, test it out and make sure the code still works as expected. Thanks!

lukewarlow replied:
That doesn't actually do formatting the output ends up being %7.5f.

I've updated for now to using a ternary to decide which g_strdup_printf to use, so only one should get allocated.

One alternative if we don't like that is to use the same approach as AXCoreObject


```c++
auto color = convertColor<SRGBA<float>>(colorValue()).resolved();
auto channel = [](float number) {
    return FormattedNumber::fixedPrecision(number, 6, TrailingZerosPolicy::Keep);
};
return color.alpha == 1
    ? makeString("rgb "_s, channel(color.red), ' ', channel(color.green), ' ', channel(color.blue), " 1"_s)
    : makeString("rgb "_s, channel(color.red), ' ', channel(color.green), ' ', channel(color.blue), ' ', channel(color.alpha));
```

The 6 would need changing to a 5 to maintain GTK's existing accuracy of 5 decimal places.

philn commented on Source/WebKit/UIProcess/API/gtk/WebKitColorChooserRequest.cpp:
```diff
@@ -134,6 +139,19 @@ static void webkit_color_chooser_request_class_init(WebKitColorChooserRequestCla
+     *
+     * The #gboolean allow-alpha of the request
+     *
+     * Since: 2.8
```
2.50

philn commented on Source/WebKit/UIProcess/API/gtk/WebKitColorChooserRequest.cpp:
```diff
@@ -196,6 +214,21 @@ void webkit_color_chooser_request_get_rgba(WebKitColorChooserRequest* request, G
+ *
+ * Returns: %TRUE if the request allows alpha, %FALSE otherwise
+ *
+ * Since: 2.8
```
Ditto

lukewarlow committed:
[GTK] Enhanced <input type=color>: support alpha picker
https://bugs.webkit.org/show_bug.cgi?id=290266

Reviewed by NOBODY (OOPS!).

Enable the GTK color picker to show its alpha slider when
the enhancements preference is enabled and the alpha attribute is used.

Note previously GTK color pickers always showed an alpha slider which never affected
the actual value, this patch also fixes this behaviour.

This adds a new API method webkit_color_chooser_request_get_allow_alpha
so that embedders can choose to update their UI accordingly.

* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/gtk/accessibility/color-well-expected.txt: Added.
* LayoutTests/platform/wpe/TestExpectations:
* Source/WebCore/accessibility/atspi/AccessibilityObjectTextAtspi.cpp:
(WebCore::AccessibilityObjectAtspi::text const):
* Source/WebKit/UIProcess/API/gtk/PageClientImpl.cpp:
(WebKit::PageClientImpl::createColorPicker):
* Source/WebKit/UIProcess/API/gtk/WebKitColorChooser.cpp:
(WebKit::WebKitColorChooser::create):
(WebKit::WebKitColorChooser::WebKitColorChooser):
* Source/WebKit/UIProcess/API/gtk/WebKitColorChooser.h:
* Source/WebKit/UIProcess/API/gtk/WebKitColorChooserRequest.cpp:
(webkitColorChooserRequestGetProperty):
(webkit_color_chooser_request_class_init):
(webkit_color_chooser_request_get_allow_alpha):
(webkitColorChooserRequestCreate):
* Source/WebKit/UIProcess/API/gtk/WebKitColorChooserRequest.h.in:
* Source/WebKit/UIProcess/gtk/WebColorPickerGtk.cpp:
(WebKit::WebColorPickerGtk::create):
(WebKit::WebColorPickerGtk::WebColorPickerGtk):
(WebKit::WebColorPickerGtk::showColorPicker):
* Source/WebKit/UIProcess/gtk/WebColorPickerGtk.h:
(WebKit::WebColorPickerGtk::supportsAlpha const):
* Tools/MiniBrowser/gtk/BrowserTab.c:
(runColorChooserCallback):

lukewarlow commented on Source/WebCore/accessibility/atspi/AccessibilityObjectTextAtspi.cpp:
```diff
@@ -276,8 +276,11 @@ String AccessibilityObjectAtspi::text() const
+        auto color = convertColor<SRGBA<float>>(m_coreObjeWct->colorValue()).resolved();
+        GUniquePtr<char> colorString(color.alpha == 1.0f
+            ? g_strdup_printf("rgb %7.5f %7.5f %7.5f 1", color.red, color.green, color.blue)
+            : g_strdup_printf("rgb %7.5f %7.5f %7.5f %7.5f", color.red, color.green, color.blue, color.alpha)
```
The linter doesn't seem to like the indentation on the second branch of the ternary (line 282) but I'm not sure why? It seems pretty standard?

TingPing replied:
I agree this seems fine and a false-positive.

Maybe put the `);` on the same line though.
